### PR TITLE
host-level ttys

### LINKED ISSUE #1016: Add terminals on the host.

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1016

### CODE REVIEW

**Comment by tomwilkie** [comment]
@2opremio would you mind squashing into 2 changesets - one for the vendoring and one for the rest?

File: probe/host/controls.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+// Control IDs used by the host integration.
+const (
+	ExecHost = "host_exec"
+)
+
+// Controls handles controls for a hosts.
+type Controls struct {
+	pipes controls.PipeClient
+}
```
Comment (tomwilkie) [inline probe/host/controls.go]: I'd prefer to see this rolled into the host Reporter.

File: common/xfer/pipes.go [hunk 2]
```diff
@@ -27,7 +27,16 @@ type pipe struct {
 	onClose         func()
 }
 
-// NewPipe makes a new... pipe.
+// NewPipeFromEnds makes a new pipe specifying its ends
+func NewPipeFromEnds(local io.ReadWriter, remote io.ReadWriter) Pipe {
```
Comment (tomwilkie) [inline common/xfer/pipes.go]: Only invocation of this has local set to `nil`.

I think I'd prefer to see a new type (that implements Pipe) but panics on `CopyToWebsocket` or something.

Comment (tomwilkie) [inline common/xfer/pipes.go]: (what I'm getting at: this type is already used in too many places for too many things, and needs to be broken down into multiple types for different uses)

**Comment by 2opremio** [comment]
> would you mind squashing into 2 changesets - one for the vendoring and one for the rest?

Done + rebased

File: probe/host/controls.go (hunk 1 above)
Comment (2opremio) [inline probe/host/controls.go]: I thought about it after you suggested offline but:

* I am planning to extend host controls with logs next, so we will probably need to separate it from the reporter anyhow.
* Now there are architecture specific files for controls.

File: common/xfer/pipes.go (hunk 2 above)
Comment (2opremio) [inline common/xfer/pipes.go]: > needs to be broken down into multiple types for different uses

AFAICT pipes are always created and passed around as a `Pipe` interface so I am not sure what's the benefit of creating new types if they al implement `Pipe`.

File: probe/host/controls.go (hunk 1 above)
Comment (tomwilkie) [inline probe/host/controls.go]: I'd keep the separate file; its the one-field struct and the extra stuff that puts in probe's main that I object to.  And for consistency with the way docker/controls.go is done.

Comment (2opremio) [inline probe/host/controls.go]: OK, will do.

Comment (2opremio) [inline probe/host/controls.go]: Actually, in docker, the controls are associated with the registry, which doesn't have a `host` equivalent. So, there's no way to make it consistent anyways.

File: common/xfer/pipes.go (hunk 2 above)
Comment (2opremio) [inline common/xfer/pipes.go]: After our conversation on Slack I now understand this. I need to think carefully about how to resolve this though. I am still not closely familiar with the full pipes though and I think this refactoring falls of out of the scope of this PR. I have created

File: probe/host/controls.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+			return
+		}
+		log.Info("Host shell closed.")
+	})
+	go func() {
+		if err := cmd.Wait(); err != nil {
+			log.Errorf("Error waiting on host shell: %v", err)
+		}
+		ptyPipe.Close()
+		pipe.Close()
```
Comment (paulbellamy) [inline probe/host/controls.go]: Is it alright to close the pipe multiple times here?

File: probe/host/controls_darwin.go [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package host
+
+var hostShellCmd = []string{"/bin/bash"}
```
Comment (paulbellamy) [inline probe/host/controls_darwin.go]: Seems like it would be better to ask the host what the user's `$SHELL` is.

**Comment by paulbellamy** [comment]
I'm *really* not convinced this is a good idea, for security reasons.

**Comment by 2opremio** [comment]
> I'm really not convinced this is a good idea, for security reasons.

This doesn't degrade security in any way. To prove it, just open a terminal in the Scope container and type `nsenter -t1 -m bash`

Productizing this makes it more evident as @rade pointed out offline, but that's it.

File: probe/host/controls.go (hunk 3 above)
Comment (2opremio) [inline probe/host/controls.go]: Those `Closes()`s do different things (`pipe.Close()` calls `OnClose()` but doesn't close the `ptyPipe`) the pattern may be a bit odd but that's how the `Pipe` interface works (something to take into consideration for #1215)

Comment (paulbellamy) [inline probe/host/controls.go]: Ok, and if the ptyPipe exits, then cmd.Wait will return, and ptyPipe.Close, and pipe.Close will be called, so I'm just checking ptyPipe.Close is safe to close multiple times, etc... The cleanup seems quite convoluted.

Comment (2opremio) [inline probe/host/controls.go]: Would adding `ptyPipe.Close()` to `OnClose()` make it easier to understand?

Comment (2opremio) [inline probe/host/controls.go]: I moved `ptyPipe.Close()` to a defer right after creating it. Regarding the closing pattern, I just followed what was done in the docker controls.

**Comment by paulbellamy** [comment]
Technically, looks fine to me. But I'm really not convinced it's a good idea.

**Comment by pidster** [comment]
LGTM
Security in Scope is an issue should be addressed separately, as there's lots more things to consider than just this.

**Pull request merged**